File: algorithms/data_structures/trie.js
// =====================================================================
// Trie / Префиксное дерево (бор)
// ---------------------------------------------------------------------
// Дерево, где каждый узел = символ. Путь от корня = строка.
// Операции (L = длина слова):
//   insert        — O(L)
//   search        — O(L)
//   startsWith    — O(L)
//   delete        — O(L)
//
// Применение: автодополнение, T9, словари, IP routing, проверка орфографии.
// =====================================================================

class TrieNode {
    constructor() {
        this.children = new Map();
        this.isEnd = false; // помечает конец слова
    }
}

class Trie {
    constructor() {
        this.root = new TrieNode();
    }

    insert(word) {
        let node = this.root;
        for (const ch of word) {
            if (!node.children.has(ch)) {
                node.children.set(ch, new TrieNode());
            }
            node = node.children.get(ch);
        }
        node.isEnd = true;
    }

    search(word) {
        const node = this._traverse(word);
        return node !== null && node.isEnd;
    }

    startsWith(prefix) {
        return this._traverse(prefix) !== null;
    }

    _traverse(s) {
        let node = this.root;
        for (const ch of s) {
            if (!node.children.has(ch)) return null;
            node = node.children.get(ch);
        }
        return node;
    }

    /**
     * Находит все слова с данным префиксом.
     */
    autocomplete(prefix) {
        const node = this._traverse(prefix);
        if (!node) return [];

        const results = [];
        this._collect(node, prefix, results);
        return results;
    }

    _collect(node, path, results) {
        if (node.isEnd) results.push(path);
        for (const [ch, child] of node.children) {
            this._collect(child, path + ch, results);
        }
    }
}

const trie = new Trie();
['apple', 'app', 'apricot', 'banana'].forEach(w => trie.insert(w));
console.log(trie.search('app')); // true
console.log(trie.search('appl')); // false (не полное слово)
console.log(trie.startsWith('appl')); // true
console.log(trie.autocomplete('ap')); // ['apple', 'app', 'apricot']

module.exports = { Trie, TrieNode };

File: algorithms/data_structures/trie.py
"""
=====================================================================
Trie / Префиксное дерево
---------------------------------------------------------------------
insert/search/startsWith — O(L)
Применение: автодополнение, словари, T9.
=====================================================================
"""
from typing import List


class TrieNode:
    __slots__ = ("children", "is_end")

    def __init__(self):
        self.children = {}
        self.is_end = False


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        node = self.root
        for ch in word:
            if ch not in node.children:
                node.children[ch] = TrieNode()
            node = node.children[ch]
        node.is_end = True

    def search(self, word: str) -> bool:
        node = self._traverse(word)
        return node is not None and node.is_end

    def starts_with(self, prefix: str) -> bool:
        return self._traverse(prefix) is not None

    def _traverse(self, s: str):
        node = self.root
        for ch in s:
            if ch not in node.children:
                return None
            node = node.children[ch]
        return node

    def autocomplete(self, prefix: str) -> List[str]:
        node = self._traverse(prefix)
        if node is None:
            return []
        results = []
        self._collect(node, prefix, results)
        return results

    def _collect(self, node: TrieNode, path: str, results: List[str]) -> None:
        if node.is_end:
            results.append(path)
        for ch, child in node.children.items():
            self._collect(child, path + ch, results)


if __name__ == "__main__":
    t = Trie()
    for w in ["apple", "app", "apricot", "banana"]:
        t.insert(w)
    print(t.search("app"))         # True
    print(t.search("appl"))        # False
    print(t.starts_with("appl"))   # True
    print(t.autocomplete("ap"))    # ['app', 'apple', 'apricot']
